### PR TITLE
SNOW-900112 fix table sproc register from file

### LINKED ISSUE #1019: SNOW-900112: register_from_file() does not work with tabular stored procedures
Simply said, register_from_file does not work for tabular stored procedures.
Please find very small reproducable example attached.

```
from snowflake.snowpark.session import Session
from snowflake.snowpark.dataframe import DataFrame
# This works
@sproc(packages=['snowflake-snowpark-python'])
def test_sproc(session: Session) -> DataFrame:
    return session.range(5)

test_sproc().show()

# This results in error: "Pandas stored procedure is not supported"
# content of sproc_code.py:
# --------------
# from snowflake.snowpark.session import Session
# from snowflake.snowpark.dataframe import DataFrame
#
# def test_sproc(session: Session) -> DataFrame:
#     return session.range(5)
session.sproc.register_from_file(
    file_path='sproc_code.py', 
    func_name='test_sproc',
    packages=['snowflake-snowpark-python']
)
```

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1019 

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Fix type hint detection when registering table sprocs using `register_from_file`.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 1]
```diff
@@ -571,7 +589,7 @@ def retrieve_func_type_hints_from_source(
 ) -> Optional[Dict[str, str]]:
     """
     Retrieve type hints of a function from a source file, or a source string (test only).
-    Returna None if the function is not found.
+    Return None if the function is not found.
     """
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/type_utils.py:593]:
```suggestion
    Returns None if the function is not found.
    """
```

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 2]
```diff
@@ -406,7 +407,7 @@ def merge_type(a: DataType, b: DataType, name: Optional[str] = None) -> DataType
         return a
 
 
-def python_type_str_to_object(tp_str: str) -> Type:
+def python_type_str_to_object(tp_str: str, object_type: str) -> Type:
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/type_utils.py:410]: Should we assert on the expected `object_type`?

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/type_utils.py:410]: Basically checking if `object_type` is of the expected types (I imagine only UDXF and sp). Are there other use cases that might call this?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:410]: We also call it with `python_type_to_snow_type` in `src/snowflake/snowpark/dataframe_na_functions.py` which is none of the objects in `TempObjectType` enums. Right now, I'm calling this function with `object_type=""`

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/type_utils.py:410]: Another option is to just pass in `is_return_type_of_sproc` and default to `False`, and if we need more than `sproc`, we could refactor it to take in `object_type` 🤔 . Will this make it clearer?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:410]: sounds like a better alternative. will work on it.

File: tests/unit/test_types.py [hunk 3]
```diff
@@ -549,7 +564,7 @@ def check_type(python_type, snow_type, is_nullable, type_str_override=None):
 
     # invalid type str
     with pytest.raises(NameError):
-        python_type_to_snow_type("string")
+        python_type_to_snow_type("string", "")
```
Comment (sfc-gh-sfan) [inline tests/unit/test_types.py:567]: Do we still need this?

**Review by sfc-gh-sfan — APPROVED** [review]
We might even be able to use callstack to deduce if we are in a stored proc creation....but i'm not sure how stable that is so we could keep things as is.

File: tests/unit/test_types.py (hunk 3 above)
Comment (sfc-gh-aalam) [reply tests/unit/test_types.py:567]: no, must have missed it in the cleanup

**Pull request merged**